### PR TITLE
feat(recipes): add evo2_megatron_quant — INT8 weight compression + ModelOpt PTQ harness

### PR DESCRIPTION
**Pull request by huthvincent**
### Description

Adds a new inference-side recipe, `recipes/evo2_megatron_quant/`, for BioNeMo biological foundation models (ESM-2, Geneformer, Evo2). Upstream already covers low-precision **training** (TE FP8/MXFP8/NVFP4); this adds an **inference** path:

- **Real INT8 weight compression** — a drop-in `nn.Linear` replacement (`INT8Linear`, per-channel symmetric) that stores weights in int8 and dequantizes on the fly. Measured on Evo2 7B: **~15% GPU-memory reduction, cosine sim ≈ 0.998, top-1 100%** vs BF16.
- **ModelOpt PTQ harness** — bridges BioNeMo model loading to `mtq.quantize()` with an automated quality report (cosine sim / top-k agreement / MSE).
- **CPU-only L0 sanity test** (`tests/test_compressed_linear_roundtrip.py`) — no GPU or checkpoint required, so it runs in fast CI.

This is a **Draft** opened to get maintainer direction on the questions in #1681:
1. Do you want an inference PTQ / weight-compression recipe in this repo, or should low-precision inference go through NeMo + TensorRT-LLM?
2. Is `recipes/evo2_megatron_quant/` the right home, or should this fold into `recipes/evo2_megatron/`?
3. I intentionally scoped this first PR to the parts that hold up (INT8 + PTQ harness). FP8/INT4 real weight compression and KV-cache experiments are held back for a follow-up once their GPU quality is re-measured.

> Note: the INT8 quantization math is verified numerically and the CPU L0 test is ready, but I have not yet executed the pytest in a torch environment — it should run under CI once authorized. Happy to align file layout (e.g. flatten `src/`), pin `requirements.txt` to the target container, and add a Dockerfile/hydra config to match your recipe conventions.

Follows up on #1681.

#### Usage

```python
from src.compress_model import compress_model

stats = compress_model(evo2_model, precision="int8")  # swaps Linear -> INT8Linear
# -> ~15% GPU memory reduction on Evo2 7B, cosine sim ~0.998 vs BF16
```

### Type of changes

- [x] New feature (non-breaking change which adds functionality)

### CI Pipeline Configuration

Only the CPU-only L0 sanity test is intended for the default PR pipeline. The GPU benchmark and `tests/test_single_model.py` need NGC checkpoints and are meant to be run manually.

### Pre-submit Checklist

- [ ] I have tested these changes locally (INT8 math verified numerically; pytest pending a torch/CI run)
- [x] I have updated the documentation accordingly (recipe README)
- [x] I have added/updated tests as needed (CPU-only L0 sanity test)
- [ ] All existing tests pass successfully (pending CI `/ok to test`)

## Summary by CodeRabbit

* **New Features**
  * Added an INT8 weight-compression quantization recipe for Evo2, ESM-2, and Geneformer using ModelOpt methods.
  * Introduced a benchmarking workflow that measures latency, GPU memory savings, and logit quality, and exports results to CSV.
  * Added container startup and pretrained model download scripts.
  * Added quantization configuration definitions covering multiple precision families (FP8/INT8/INT4/NVFP4/microscaling/mixed).

* **Documentation**
  * Added recipe documentation with setup, supported checkpoints, and quick-start/benchmark guidance.

* **Tests**
  * Added CPU-only INT8 linear roundtrip tests and a single-model quantization validation script.

### CODE REVIEW

**Comment by huthvincent** [comment]
@coderabbitai review

**Comment by huthvincent** [comment]
Thanks for the thorough review! Addressed the findings in `46eb3bde`:

**Fixed**
- `tests/test_single_model.py` — renamed `test_single_method` → `run_single_method` so pytest won't collect it as a test; `main()` now returns a non-zero exit code when any method FAILs/ERRORs.
- `src/compress_model.py` — switched to a package-relative import (`from .compressed_linear import ...`) with a top-level fallback for the `sys.path`-insert case.
- `src/quantize.py` — calibration forward-pass failures are now logged (with a warning if *all* passes fail) instead of being silently swallowed; dropped the always-zero `num_quantizers` field.
- `scripts/benchmark_real_quant.py` — exception-safe sequence-length sweep (vars initialized before `try`, cleaned up in `finally`), free the warmup output, `--max-seqlen` is now disableable (`BooleanOptionalAction` → `--no-max-seqlen`), stale result file removed before each worker run, `TimeoutExpired` handled, and stopped labeling the linear sweep a "binary search".
- `docker/start_container.sh` — default mount is now the recipe root (was an empty `./quantization` subdir); `--net=host` is opt-in via `USE_HOST_NET=1`.
- `scripts/download_models.sh` — unknown model names now fail fast with a non-zero exit.
- `src/__init__.py` — dropped AMPLIFY from the supported-models docstring (only ESM-2/Geneformer/Evo2 are registered).
- `README.md` — language-tagged the layout fence (MD040); `requirements.txt` — pinned `pyyaml`/`pytest` (torch and modelopt come from the BioNeMo base image, so they're intentionally not re-pinned).

**One I believe is a false positive**
- `tests/test_compressed_linear_roundtrip.py` (cosine test): `return_bias=False` does **not** drop the bias from the computation — `INT8Linear.forward` always applies it via `F.linear(x, w, self.bias)`; `return_bias` only controls whether `forward` returns `(out, None)` vs a bare tensor. So the test compares equivalent paths. Happy to add an explicit bias-enabled case as well if you'd prefer.

**Comment by huthvincent** [comment]
@coderabbitai review

**Comment by huthvincent** [comment]
Second pass addressed in `6f0f5e48`:

**Fixed as suggested**
- `scripts/benchmark_real_quant.py` — `json.load` is now guarded against a truncated worker result: a config whose worker was killed mid-write returns a `CRASH` row instead of raising `JSONDecodeError` out of `run_test()` and aborting the whole sweep before any CSV is written. Good catch.
- `src/quantize.py` — `--quant` tokens are stripped, and the resolved name is whitelisted against `ALL_QUANT_METHODS` before it reaches `getattr(mtq, ...)`.
- `src/quantize.py`, `src/adapters.py`, `scripts/benchmark_real_quant.py` — replaced the deprecated `torch.cuda.amp.autocast` with `torch.autocast(device_type="cuda", ...)`.

**Two I solved differently — reasoning below, happy to revisit**

1. **`quant_cfg` list layout (`quantize.py`).** I agree with the risk you identified — a silent no-op for `--all-mlp` is the real danger here. But I couldn't verify the "ModelOpt 0.45+ stores `quant_cfg` as an ordered list" claim from my environment, and writing speculative code against an unverified internal layout seemed worse than the bug itself. So instead of guessing at the list shape, I made the failure **loud**: if `quant_cfg` isn't a mapping containing a `default` entry, it now raises with the offending type rather than silently doing nothing. For a measurement harness, erroring out beats quietly reporting default-scope numbers as "all-MLP". If you can point me at the ModelOpt version and the concrete entry shape, I'll happily add explicit list handling.

2. **Focused tests for the exit-code contract (`test_single_model.py`).** The root problem was that the file was misnamed: it is a **manual GPU runner** (needs a CUDA device and an NGC checkpoint), not a pytest test — and its `tests/test_*.py` name is exactly what made pytest try to collect it in your first review. Rather than bolt mocks onto a GPU harness to make it look like a unit test, I moved it to **`scripts/run_single_model.py`**. `tests/` now contains only the genuine CPU-only L0 gate (`test_compressed_linear_roundtrip.py`), which needs no GPU or checkpoint and is what CI should actually run. That removes the collection hazard at the root rather than papering over it.
